src/build/env.py

```diff
@@ -51,191 +24,166 @@ class IsolatedEnvironment(object):
     Non-standard paths injected directly to sys.path still be passed to the environment.
     """
 
-    _MANIPULATE_PATHS = ('purelib', 'platlib')
+    def __init__(self, path, executable, install_executable):
+        # type: (str, str, str) -> None
+        """
+        Define an isolated build environment.
 
-    def __init__(self, remove_paths, _executable=sys.executable):  # type: (Sequence[str], str) -> None
+        :param path: the path where the environment exists
+        :param executable: the python executable within the environment
+        :param install_executable: an executable that allows installing packages within the environment
         """
-        :param remove_paths: Import paths that should be removed from the environment
+        self._path = path
+        self._install_executable = install_executable
+        self._executable = executable
+
+    @classmethod
+    def for_current(cls):  # type: () -> IsolatedEnvironment
+        """
+        Create an isolated build environment into a temporary folder that matches the current interpreter.
+
+        :return: the isolated build environment
         """
-        self._env = {}  # type: Dict[str, Optional[str]]
-        self._env_vars = {}  # type: Dict[str, str]
-        self._path = None  # type: Optional[str]
-        self._remove_paths = []  # type: List[str]
-        self._executable = _executable
-        self._old_executable = None  # type: Optional[str]
-        self._pip_executable = None  # type: Optional[str]
-
-        # normalize paths so that we can compare them -- required on case insensitive systems
-        for path in remove_paths:
-            self._remove_paths.append(os.path.normcase(path))
+        path = tempfile.mkdtemp(prefix='build-env-')
+        executable, pip_executable = _create_isolated_env(path)
+        return cls(path=path, executable=executable, install_executable=pip_executable)
+
+    def __enter__(self):  # type: () -> IsolatedEnvironment
+        """Enable the isolated build environment"""
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        # type: (Optional[Type[BaseException]], Optional[BaseException], Optional[TracebackType]) -> None
+        """
+        Exit from the isolated build environment.
+
+        :param exc_type: the type of exception raised (if any)
+        :param exc_val: the value of exception raised (if any)
+        :param exc_tb: the traceback of exception raised (if any)
+        """
+        self.close()
```

Comment:
Users who want to reuse an isolated build environment multiple times should not invoke this class's context manager. However, they must invoke ``close`` at the end of their application to trigger the cleanup.

Reply:
I don't think we should call `close` here. This should deactivate the environment, not close it.

We should call `close` instead in `__del__`. I am not even sure if there is need for an explicit destructor. IMO we should just rename `close` to `__del__`.

Reply:
That's very much against the language design and guarantees. Relying on ``__del__`` to be called is just not supported/guaranteed within the Python language. See here https://bugs.python.org/issue39360 for more information about a nasty bug this can cause, and where core developers express that in Python you should always explicitly cleanup and never rely on ``__del__`` to be called for resources.

Reply:
If you want to explicitly clean up the temporary directory, you can call the destructor. But you are not necessarily required to. We are not relying on `__del__` to be called for anything.
If you want to keep `env.close()` as an alternative to `del env`, sure.

The main point here is that I do not think we should be forcefully cleaning up the environment when we exit. Maybe we can pass an argument to opt in/out?

Reply:
I think the only thing you should be cleaning up within a ``__del__`` method is memory. Not folders and files.

> The main point here is that I do not think we should be forcefully cleaning up the environment when we exit

Why? We created it just to be able to perform the build. Why would we not clean it up? At least within build. If we're talking about some custom application they can always use the API and not call the close /context amanger if they want to reuse across runs.